ntpsnmpd/ntpsnmpd.1
.TH NTPSNMPD 1 2011-12-08 "( 4.2.6p5-RC3)" "Programmer's Manual"
.\"  EDIT THIS FILE WITH CAUTION  (ntpsnmpd.1)
.\"  
.\"  It has been AutoGen-ed  December  8, 2011 at 09:08:46 PM by AutoGen 5.12
.\"  From the definitions    ntpsnmpd-opts.def
.\"  and the template file   agman1.tpl
.\"
.SH NAME
ntpsnmpd \- NTP SNMP MIB agent
.SH SYNOPSIS
.B ntpsnmpd
.\" Mixture of short (flag) options and long options
.RB [ \-\fIflag\fP " [\fIvalue\fP]]... [" \--\fIopt-name\fP " [[=| ]\fIvalue\fP]]..."
.PP
All arguments must be options.
.SH "DESCRIPTION"
This manual page briefly documents the \fBntpsnmpd\fP command.
Its description is not documented.
.SH OPTIONS
.TP
.BR \-n ", " \--nofork
Do not fork.
.sp

.TP
.BR \-p ", " \--syslog
Log to syslog().
.sp

.TP
.BR \--agentxsocket "=\fIstring\fP"
The socket address ntpsnmpd uses to connect to net-snmpd.
The default \fIstring\fP for this option is:
.ti +4
 unix:/var/agentx/master
.sp
[<transport-specifier>:]<transport-address>
The default is the Unix Domain socket "unix:/var/agentx/master". Another common alternative is tcp:localhost:705.
.TP
.BR \-? , " \--help"
Display extended usage information and exit.
.TP
.BR \-! , " \--more-help"
Extended usage information passed thru pager.
.TP
.BR \-> " [\fIrcfile\fP]," " \--save-opts" "[=\fIrcfile\fP]"
Save the option state to \fIrcfile\fP.  The default is the \fIlast\fP
configuration file listed in the \fBOPTION PRESETS\fP section, below.
.TP
.BR \-< " \fIrcfile\fP," " \--load-opts" "=\fIrcfile\fP," " \--no-load-opts"
Load options from \fIrcfile\fP.
The \fIno-load-opts\fP form will disable the loading
of earlier RC/INI files.  \fI--no-load-opts\fP is handled early,
out of order.
.TP
.BR \- " [{\fIv|c|n\fP}]," " \--version" "[=\fI{v|c|n}\fP]"
Output version of program and exit.  The default mode is `v', a simple
version.  The `c' mode will print copyright information and `n' will
print the full copyright notice.
.SH OPTION PRESETS
Any option that is not marked as \fInot presettable\fP may be preset
by loading values from configuration ("RC" or ".INI") file(s) and values from
environment variables named:
.nf
  \fBNTPSNMPD_<option-name>\fP or \fBNTPSNMPD\fP
.fi
.ad
The environmental presets take precedence (are processed later than)
the configuration files.
The \fIhomerc\fP files are "\fI$HOME\fP", and "\fI.\fP".
If any of these are directories, then the file \fI.ntprc\fP
is searched for within those directories.
.SH AUTHOR
David L. Mills and/or others
.br
Please send bug reports to:  http://bugs.ntp.org, bugs@ntp.org

.PP
.nf
.na
see html/copyright.html

.fi
.ad
.PP
This manual page was \fIAutoGen\fP-erated from the \fBntpsnmpd\fP
option definitions.


NEWS
---
NTP 4.2.8-

Important Changes

* Internal NTP Era counters

The internal counters that track which "era" (range of years) we are in
rolls over every 136 years'.  The current "era" started at the stroke of
midnight on 1 Jan 1900, and ends just before the stroke of midnight on
1 Jan 2036.
In the past, we have used the "midpoint" of the  range to decide which
era we were in.  Given the longevity of some products, it became clear
that it would be more functional to "look back" less, and "look forward"
more.  We now compile a timestamp into the ntpd executable and when we
get a timestamp we us the "built-on" to tell us what era we are in.
This check "looks back" 10 years, and "looks forward" 126 years.

So if you have a system that ...

* ntpdc responses disabled by default

Dave Hart writes:

For a long time, ntpq and its mostly text-based mode 6 (control) 
protocol have been preferred over ntpdc and its mode 7 (private 
request) protocol for runtime queries and configuration.  There has 
been a goal of deprecating ntpdc, previously held back by numerous 
capabilities exposed by ntpdc with no ntpq equivalent.  I have been 
adding commands to ntpq to cover these cases, and I believe I've 
covered them all, though I've not compared command-by-command 
recently. 

As I've said previously, the binary mode 7 protocol involves a lot of 
hand-rolled structure layout and byte-swapping code in both ntpd and 
ntpdc which is hard to get right.  As ntpd grows and changes, the 
changes are difficult to expose via ntpdc while maintaining forward 
and backward compatibility between ntpdc and ntpd.  In contrast, 
ntpq's text-based, label=value approach involves more code reuse and 
allows compatible changes without extra work in most cases. 

Mode 7 has always been defined as vendor/implementation-specific while 
mode 6 is described in RFC 1305 and intended to be open to interop 
with other implementations.  There is an early draft of an updated 
mode 6 description that likely will join the other NTPv4 RFCs 
eventually. (http://tools.ietf.org/html/draft-odonoghue-ntpv4-control-01)

For these reasons, ntpd 4.2.7p230 by default disables processing of 
ntpdc queries, reducing ntpd's attack surface and functionally 
deprecating ntpdc.  If you are in the habit of using ntpdc for certain 
operations, please try the ntpq equivalent.  If there's no equivalent, 
please open a bug report at http://bugs.ntp.org./

--- 
NTP 4.2.6p5-RC3 (Harlan Stenn <stenn@ntp.org>, 2011/12/xx) 
 
Focus: Bug fix
 
Severity: Medium 
 
This is a recommended upgrade. 

This release includes a fix for a mismatch problem with IP addresses
specified in "nic" or "interface" directives in the ntpd configuration
file using the address/prefix format, such as "fe80::/64" erroneously
matching localhost [::1] and wildcard [::].  It also includes fixes
for orphan mode, extraneous creation of ephemeral associations from
certain broadcastclient and multicastclient configurations, and some
display issues with ntpq.

ntpd

 * Correctly match IP address prefixes for "nic" and "interface".
 * Fix orphan mode stratum incorrectly counting to infinity.
 * Orphan parent selection metric was missing ntohl().
 * No longer send nonprintable stratum 16 refid to ntpq.
 * broadcastclient, multicastclient would spin up duplicate
   ephemeral associations without broadcastdelay.
 * Exclude not-yet-determined sys_refid from use in loopback TEST12.
 * Never send KoD rate limiting response to MODE_SERVER response.
 * Fix 3-char refid sent by ntpd 4.2.6p5-RC2 that ends with extra dot.
 * clock_update() sys_rootdisp calculation should include root delay.
 * get_systime() should not offset by sys_residual.
 * sys_jitter calculation overweights sys.peer jitter.
 * Ensure NULL peer->dstadr is not accessed in orphan parent selection.

ntpq

 * Optional billboard column "server" did not honor -n.
 * lopeers ipv6 "local" used to overrun its column.

--- 
NTP 4.2.6p4 (Harlan Stenn <stenn@ntp.org>, 2011/09/22) 
 
Focus: Bug fixes and portability improvements 
 
Severity: Medium 
 
This is a recommended upgrade. 
 
This release includes build infrastructure updates, code 
clean-ups, minor bug fixes, fixes for a number of minor 
ref-clock issues, and documentation revisions. 
 
Portability improvements affect AIX, HP-UX, Linux, OS X and 64-bit time_t. 
 
New features / changes in this release: 
 
Build system 
 
* Fix checking for struct rtattr 
* Update config.guess and config.sub for AIX 
* Upgrade required version of autogen and libopts for building 
  from our source code repository 
 
ntpd 
 
* Backported several fixes for Coverity warnings from ntp-dev 
* Fix a rare boundry condition in UNLINK_EXPR_SLIST() 
* Allow "logconfig =allall" configuration directive 
* Bind tentative IPv6 addresses on Linux 
* Correct WWVB/Spectracom driver to timestamp CR instead of LF 
* Improved tally bit handling to prevent incorrect ntpq peer status reports 
* Exclude the Undisciplined Local Clock and ACTS drivers from the initial 
  candidate list unless they are designated a "prefer peer" 
* Prevent the consideration of Undisciplined Local Clock or ACTS drivers for 
  selection during the 'tos orphanwait' period 
* Prefer an Orphan Mode Parent over the Undisciplined Local Clock or ACTS 
  drivers 
* Improved support of the Parse Refclock trusttime flag in Meinberg mode 
* Backport utility routines from ntp-dev: mprintf(), emalloc_zero() 
* Added the NTPD_TICKADJ_PPM environment variable for specifying baseline 
  clock slew on Microsoft Windows 
* Code cleanup in libntpq 
 
ntpdc 
 
* Fix timerstats reporting 
 
ntpdate 
 
* Reduce time required to set clock 
* Allow a timeout greater than 2 seconds 
 
sntp 
 
* Backward incompatible command-line option change: 
  -l/--filelog changed -l/--logfile (to be consistent with ntpd) 
 
Documentation 
 
* Update html2man. Fix some tags in the .html files 
* Distribute ntp-wait.html 

---
NTP 4.2.6p3 (Harlan Stenn <stenn@ntp.org>, 2011/01/03)


ntpsnmpd/ntpsnmpd.c
/*****************************************************************************
 *
 *  ntpsnmpd.c
 *
 *  The NTP SNMP daemon is an Agent X subagent application that 
 *  registers itself with a running SNMP Agent X master process running
 *  on the same machine on port TCP 705. It utilizes the libntqp library
 *  which accesses status and general data of a running ntpd process on
 *  the same machine and enables the user to monitor the status of a
 *  ntp daemon process via SNMP. 
 *
 *  This started as a Google Summer of Code 2008 project, 
 *  including the ntpsnmpd sub agent and the libntpq library. 
 *
 *  For more information please visit
 *	http://support.ntp.org/bin/view/Dev/GSoC2008snmp
 *  Or contact:
 *   Harlan Stenn   (Mentor) at stenn@ntp.org
 *   Heiko Gerstung (Student) at gerstung@ntp.org 
 *
 ****************************************************************************/

#include <ntp_snmp.h>
#include <signal.h>
#include <sys/time.h>

#ifdef SOLARIS /* needed with at least Solaris 8 */
#include <siginfo.h>
#endif

#include <libntpq.h>
#include <ntpsnmpd-opts.h>

static int keep_running;

RETSIGTYPE
stop_server(int a) {
    keep_running = 0;
}

/* The main function just sets up a few things and then enters a loop in which it will 
 * wait for SNMP requests coming from the master agent 
 */

int
main (int argc, char **argv) {
  int background = 0; /* start as background process */
  int use_syslog = 1; /* use syslog for logging */
  char varvalue[1024];
	

	{
		int optct = optionProcess(&ntpsnmpdOptions, argc, argv);
		argc -= optct;
		argv += optct;
	}

	if (!HAVE_OPT(NOFORK))
		background = 1;

	if (!HAVE_OPT(SYSLOG))
		use_syslog = 0;

  /* using the net-snmp syslog facility */
  if (use_syslog)
    snmp_enable_calllog();
  else
    snmp_enable_stderrlog();

  /* Become Subagent */
    netsnmp_ds_set_boolean(NETSNMP_DS_APPLICATION_ID, NETSNMP_DS_AGENT_ROLE, 1);
 
  /* go into background mode, if requested */
  if (background && netsnmp_daemonize(1, !use_syslog))
      exit(1);

  /* Now register with the master Agent X process */

  /* call Netsnmp socket startup macro, which will initialize the network stuff if required */
  SOCK_STARTUP;

  /* Set AgentX socket interface */
  netsnmp_ds_set_string(NETSNMP_DS_APPLICATION_ID,
                            NETSNMP_DS_AGENT_X_SOCKET, OPT_ARG( AGENTXSOCKET ));

  init_agent("ntpsnmpd");

  /* Try to connect to ntpd */
  if ( ntpq_openhost("localhost") == 0 )
  {
	fprintf(stderr, "Error: Could not connect to ntpd. Aborting.\n");
	exit(1);
  }
  

  /* Register callback functions ...  */
  init_ntpSnmpSubagentObject();  
  init_snmp("ntpsnmpd");

  /* Signal handler */
  keep_running = 1;
  signal(SIGTERM, stop_server);
  signal(SIGINT, stop_server);

  snmp_log(LOG_INFO,"ntpsnmpd started.\n");
  
  /* main loop here... */
  while(keep_running) {
	agent_check_and_process(1); /* 0 == don't block */
  }

  /* at shutdown time */
  ntpq_closehost();
  snmp_shutdown("ntpsnmpd");
  SOCK_CLEANUP;

  return 0;
}

